pkg/agent/protocol/http/proxy.go

```diff
@@ -97,56 +87,85 @@ type httpHandler struct {
 	client      httpClient
 }
 
-func (h *httpHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
-	var statusCode int
-	headers := http.Header{}
-	body := io.NopCloser(strings.NewReader(h.disruption.ErrorBody))
-
-	excluded := contains(h.disruption.Excluded, req.URL.Path)
-
-	if !excluded && h.disruption.ErrorRate > 0 && rand.Float32() <= h.disruption.ErrorRate {
-		// force error code
-		statusCode = int(h.disruption.ErrorCode)
-	} else {
-		req.Host = h.upstreamURL.Host
-		req.URL.Host = h.upstreamURL.Host
-		req.URL.Scheme = h.upstreamURL.Scheme
-		req.RequestURI = ""
-		originServerResponse, srvErr := h.client.Do(req)
-		if srvErr != nil {
-			rw.WriteHeader(http.StatusInternalServerError)
-			_, _ = fmt.Fprint(rw, srvErr)
-			return
+// isExcluded checks whether a request should be proxied through without any kind of modification whatsoever.
+func (h *httpHandler) isExcluded(r *http.Request) bool {
+	for _, excluded := range h.disruption.Excluded {
+		if strings.EqualFold(r.URL.Path, excluded) {
+			return true
 		}
+	}
 
-		headers = originServerResponse.Header
-		statusCode = originServerResponse.StatusCode
-		body = originServerResponse.Body
+	return false
+}
 
-		defer func() {
-			_ = originServerResponse.Body.Close()
-		}()
-	}
+// proxy proxies a request to the upstream URL.
+// Request is performed immediately, but response won't be copied to the client until wait has been consumed.
```

Comment:
I find confusing this wait channel. I think it obscures the purpose of the proxy which should in my opinion only do one thing. 

I suggest that instead of passing this argument, you use a wrapper function that waits before (or after) the given delay and then calls a function passed as an argument, given that both `proxy` and `fault` has the same signature.

Reply:
I don't have any off the top of my head, but at the very least, I would pass the delay and make the wait inside the functions instead of passing a channel. That makes the purpose of the argument clear.  I don't see any advantage on using a channel.